#+options: ':nil *:t -:t ::t <:t H:3 \n:nil ^:nil arch:headline author:t
#+options: broken-links:nil c:nil creator:nil d:(not "LOGBOOK") date:t e:t
#+options: email:nil f:t inline:t num:t p:nil pri:nil prop:nil stat:t tags:t
#+options: tasks:t tex:t timestamp:t title:t toc:nil todo:t |:t
#+title: Polymorphism-aware phylogenetic models
#+subtitle: Workshop, MIC-Phy 2021
#+author: Dominik Schrempf
#+email: dominik.schrempf@gmail.com
#+language: en
#+select_tags: export
#+exclude_tags: noexport
#+creator: Emacs 27.1 (Org mode 9.4.4)

#+startup: beamer
#+latex_class: myPresentation
#+latex_class_options: [aspectratio=169,minted]
#+latex_header: \addbibresource{~/Evolutionary-Biology/Bibliography/bibliography.bib}
#+latex_header_extra: \titlegraphic{\includegraphics[width=10em]{logos/elte}}
#+latex_compiler: unused; see `org-latex-pdf-process'

#+columns: %45ITEM %10BEAMER_env(Env) %10BEAMER_act(Act) %4BEAMER_col(Col) %8BEAMER_opt(Opt)
#+date: February 16, 2021
#+description:
#+keywords:
#+options: H:1

* Introduction
This workshop is [[https://github.com/pomo-dev/micphy-workshop][available on GitHub]].
#+beamer: \vspace{2ex}

Our goal is to understand how we can use polymorphism-aware phylogenetic models
(PoMo) to improve inferences from population data.
#+beamer: \vspace{2ex}

In the course of this workshop, we will infer a phylogenetic tree from test data
using IQ-TREE2.
#+beamer: \vspace{2ex}

If you need help, please interrupt me anytime!

* Preparation - Command line shell
:PROPERTIES:
:ID:       c1b7587a-de6c-4d16-92e5-722e3c089594
:END:
Basic knowledge of the command line shell of your choice is assumed.
- If you do not know basic commands such as =cd=, =ls=, or =less=, just lean
  back and listen to the presentation.
- Otherwise, try to follow the steps and complete the workshop yourself.
#+beamer: \vspace{1ex}

In case you are lost:
- Have a look at the manual pages, if they exist.
  #+name: ManLess
  #+begin_src sh :exports code :results none
  man less
  #+end_src
- Read how commands are used.
  #+name: HelpLess
  #+begin_src sh :exports code :results none
  less --help
  #+end_src
  
* Preparation - Download workshop and data
Option 1: If you have =git= installed, use it.
#+name: Workshop-Git
#+begin_src sh :exports code :results none :eval never
git clone https://github.com/pomo-dev/micphy-workshop.git
cd micphy-workshop
#+end_src

Option 2: Manually download the archive (requires =wget=, and =unzip=).
#+name: Workshop-ManualDownload
#+begin_src sh :exports code :results none :eval never
wget https://github.com/pomo-dev/micphy-workshop/archive/master.zip
unzip master.zip
cd micphy-workshop-master
#+end_src

The advantage of Option 1 is that you can:
- update your working tree if I have to change something during the workshop;
  use =git pull=;
- reset to the initial state if you mess up; use =git reset --hard HEAD= (be
  careful, this erases all changes made by you).

* Preparation - Install IQ-TREE2
Option 1: Install from the repository of your distribution. For example, use the
Arch Linux User Repository.
#+name: IqTree2-InstallFromRepository
#+begin_src sh :exports code :results none :eval never
yay -S iqtree
aura -A iqtree
#+end_src

Option 2: Compile yourself (not shown).
#+beamer: \vspace{2ex}

Option 3: Use =nix-shell= and the =shell.nix= expression provided in the base
directory of the repository (requires =nix=).

#+name: IqTree2-NixShell
#+begin_src sh :exports both :results output verbatim replace
nix-shell
#+end_src

#+latex: {\footnotesize
#+RESULTS: IqTree2-NixShell
: Welcome to the MIC-Phy PoMo workshop.
: The following version of IQ-TREE2 is available:
: IQ-TREE multicore version 2.1.2 COVID-edition for Linux 64-bit built Jan  1 1980
: Developed by Bui Quang Minh, James Barbetti, Nguyen Lam Tung,
: Olga Chernomor, Heiko Schmidt, Dominik Schrempf, Michael Woodhams.
: 
#+latex: }

* Preparation - Install IQ-TREE2

Option 4: Download the binary executable from the [[http://www.iqtree.org/#download][IQ-TREE2 homepage]].
#+name: IqTree2-ManualDownload
#+begin_src sh :exports code :eval never 
wget https://github.com/iqtree/iqtree2/releases/download/v2.1.2/iqtree-2.1.2-Linux.tar.gz
#+end_src

Make sure that you have permission to execute the file (=chmod +x=), and that
the executable is in your =PATH= (or that you provide the path during execution).
#+name: IqTree2-ManualInstall
#+begin_src sh :exports code :eval never 
tar -xzvf iqtree-2.1.2-Linux.tar.gz
chmod +x iqtree-2.1.2-Linux/bin/iqtree2  # Should not be necessary, but who knows.
mv iqtree-2.1.2-Linux/bin/iqtree2 ~/bin/ # If ~/bin is in your PATH.
#+end_src

* Preparation - Test IQ-TREE2 version
Depending on how you installed IQ-TREE2, please check that the version agrees
with the one I am using.

#+name: IqTree2-Version
#+begin_src sh :exports both :results output verbatim replace
iqtree2 --version
# /path/to/iqtree2 --version
# ./relative/path/to/iqtree2 --version
#+end_src

#+RESULTS: IqTree2-Version
: IQ-TREE multicore version 2.1.2 COVID-edition for Linux 64-bit built Jan  1 1980
: Developed by Bui Quang Minh, James Barbetti, Nguyen Lam Tung,
: Olga Chernomor, Heiko Schmidt, Dominik Schrempf, Michael Woodhams.
: 

* Exercise - IQ-TREE2: Access help
A convenient way to access the help is
#+name: IqTree2-Help
#+begin_src sh :exports code :eval never
iqtree2 --help | less
#+end_src

Here I print the first lines:
#+name: IqTree2-HelpFirstLines
#+begin_src sh :exports results :results verbatim 
iqtree2 --help | head -n 10
#+end_src

#+RESULTS: IqTree2-HelpFirstLines
#+begin_example
IQ-TREE multicore version 2.1.2 COVID-edition for Linux 64-bit built Jan  1 1980
Developed by Bui Quang Minh, James Barbetti, Nguyen Lam Tung,
Olga Chernomor, Heiko Schmidt, Dominik Schrempf, Michael Woodhams.

Usage: iqtree [-s ALIGNMENT] [-p PARTITION] [-m MODEL] [-t TREE] ...

GENERAL OPTIONS:
  -h, --help           Print (more) help usages
  -s FILE[,...,FILE]   PHYLIP/FASTA/NEXUS/CLUSTAL/MSF alignment file(s)
  -s DIR               Directory of alignment files
#+end_example

* Fruit fly data
Data from [[https://popfly.uab.cat][PopFly]][fn:1]. 9 populations with an average number of samples per
population of approximately 19, and an estimated heterozygosity of 0.0109:
- NTH :: Netherlands
- EG :: Egypt
- FR :: France
- GA :: Gabon
- GU :: Guinea
- EF :: Ethiopia
- KN :: Kenyia
- SB :: South Africa (Barkly East)
- SP :: South Africa (Phalaborwa)
 
/Why is it important to check the heterozygosity?/

* Exercise - DNA substitution model
Before running PoMo, we will use a normal DNA substitution model.

* Exercise - Run PoMo
#+name: IqTree2Run
#+begin_src sh :exports code :eval query 
iqtree2 -nt 4 -redo -s data/fruit_flies_1000.cf -m HKY+P+N9 -pre fruit_flies_1000.cf.N9
#+end_src

- Run PoMo.
- Find best N.
- Compare different DNA substitution models.
- Use gamma rate heterogeneity.
- Probably perform model test?
- Bootstrapping.
- Compare branch lengths for different N values.
- Compare to using normal DNA substitution models.

- Probably provide results for the 10k alignment (different N, different G, with
  bootstrapping?).

* Advice
Sometimes, the inference is unsuccessful. This may have several reasons. Two of
them are:
- The likelihood derivative is zero or close to zero and numerical underflow
  occurs. This is especially an issue when \(N\) is large. Try using =-safe=
  (which is slower).
- The algorithm diverges. Try repeating the analysis with a different seed.

In general, it is recommended to *perform replicate analyses* and compare the
parameters and log likelihoods. Further, a *good starting tree* can save a lot
of time and worries.

* Literature
:PROPERTIES:
:ID:       a3a4fe49-79a8-4618-bcae-655485ac54c4
:END:
#+attr_latex: :options [Advanced models with PoMo]
- PoMo :: textcite:DeMaio2015.
- Reversible PoMo :: textcite:Schrempf2016.
- Non-reversible PoMo :: textcite:Schrempf2017.
- Advanced models with PoMo :: textcite:Schrempf2019.
- IQ-TREE2 :: textcite:Minh2020a.
- Consistency of PoMo :: textcite:Borges2020.
- PoMo with selection :: textcite:Borges2019.

* Bibliography
:PROPERTIES:
:BEAMER_opt: allowframebreaks
:END:

#+begin_export latex
\printbibliography{}
#+end_export

* Footnotes

[fn:1] textcite:Hervas2017. 
