#+options: ':nil *:t -:t ::t <:t H:3 \n:nil ^:nil arch:headline author:t
#+options: broken-links:nil c:nil creator:nil d:(not "LOGBOOK") date:t e:t
#+options: email:nil f:t inline:t num:t p:nil pri:nil prop:nil stat:t tags:t
#+options: tasks:t tex:t timestamp:t title:t toc:nil todo:t |:t
#+title: Polymorphism-aware phylogenetic models
#+subtitle: Workshop, MIC-Phy 2021
#+author: Dominik Schrempf
#+email: dominik.schrempf@gmail.com
#+language: en
#+select_tags: export
#+exclude_tags: noexport
#+creator: Emacs 27.1 (Org mode 9.4.4)

#+startup: beamer
#+latex_class: myPresentation
#+latex_class_options: [aspectratio=169,minted,svg]
#+latex_header: \addbibresource{~/Evolutionary-Biology/Bibliography/bibliography.bib}
#+latex_header_extra: \titlegraphic{\includegraphics[width=10em]{logos/elte}}
#+latex_compiler: unused; see `org-latex-pdf-process'

#+columns: %45ITEM %10BEAMER_env(Env) %10BEAMER_act(Act) %4BEAMER_col(Col) %8BEAMER_opt(Opt)
#+date: February 16, 2021
#+description:
#+keywords:
#+options: H:1

#+property: header-args:ipython :session PoMo-Workshop

#+name: Imports
#+begin_src ipython :exports none :results none
import numpy as np
import pandas as pd
import scipy as sp

import matplotlib as mpl
import matplotlib.pyplot as plt
import seaborn as sb

sb.set()
sb.set_style("ticks", {"axes.grid": True})
sb.set_context("notebook")
%config InlineBackend.figure_format = 'svg'
#+end_src

* Introduction
This workshop is [[https://github.com/pomo-dev/micphy-workshop][available on GitHub]], please also have a look at the
corresponding [[file:PoMo-Presentation.pdf][workshop presentation]].
#+beamer: \vspace{2ex}

Our goal is to understand how we can use polymorphism-aware phylogenetic models
(PoMo) to improve inferences from population data.
#+beamer: \vspace{2ex}

In the course of this workshop, we will infer a phylogenetic tree from test data
using IQ-TREE2.
#+beamer: \vspace{2ex}

If you need help, please interrupt me anytime!

* Preparation - Command line shell
:PROPERTIES:
:ID:       c1b7587a-de6c-4d16-92e5-722e3c089594
:END:
Basic knowledge of the command line shell of your choice is assumed.
- If you do not know basic commands such as =cd=, =ls=, or =less=, just lean
  back and listen to the presentation.
- Otherwise, try to follow the steps and complete the workshop yourself.
#+beamer: \vspace{1ex}

In case you are lost:
- Have a look at the manual pages, if they exist.
  #+name: ManLess
  #+begin_src sh :exports code :results none
  man less
  #+end_src
- Read how commands are used.
  #+name: HelpLess
  #+begin_src sh :exports code :results none
  less --help
  #+end_src
  
* Preparation - Download workshop and data
Option 1: If you have =git= installed, use it.
#+name: Workshop-Git
#+begin_src sh :exports code :results none :eval never
git clone https://github.com/pomo-dev/micphy-workshop.git
cd micphy-workshop
#+end_src

Option 2: Manually download the archive (requires =wget=, and =unzip=).
#+name: Workshop-ManualDownload
#+begin_src sh :exports code :results none :eval never
wget https://github.com/pomo-dev/micphy-workshop/archive/master.zip
unzip master.zip
cd micphy-workshop-master
#+end_src

The advantage of Option 1 is that you can:
- update your working tree if I have to change something during the workshop;
  use =git pull=;
- reset to the initial state if you mess up; use =git reset --hard HEAD= (be
  careful, this erases all changes made by you).

* Preparation - Install IQ-TREE2
Option 1: Install from the repository of your distribution. For example, use the
Arch Linux User Repository.
#+name: IqTree2-InstallFromRepository
#+begin_src sh :exports code :results none :eval never
yay -S iqtree
aura -A iqtree
#+end_src

Option 2: Compile yourself (not shown).
#+beamer: \vspace{2ex}

Option 3: Use =nix-shell= and the =shell.nix= expression provided in the base
directory of the repository (requires =nix=).

#+name: IqTree2-NixShell
#+begin_src sh :exports both :results output verbatim replace
nix-shell
#+end_src

#+latex: {\footnotesize
#+RESULTS: IqTree2-NixShell
: Welcome to the MIC-Phy PoMo workshop.
: The following version of IQ-TREE2 is available:
: IQ-TREE multicore version 2.1.2 COVID-edition for Linux 64-bit built Jan  1 1980
: Developed by Bui Quang Minh, James Barbetti, Nguyen Lam Tung,
: Olga Chernomor, Heiko Schmidt, Dominik Schrempf, Michael Woodhams.
: 
#+latex: }

* Preparation - Install IQ-TREE2

Option 4: Download the binary executable from the [[http://www.iqtree.org/#download][IQ-TREE2 homepage]].
#+name: IqTree2-ManualDownload
#+begin_src sh :exports code :eval never 
wget https://github.com/iqtree/iqtree2/releases/download/v2.1.2/iqtree-2.1.2-Linux.tar.gz
#+end_src

Make sure that you have permission to execute the file (=chmod +x=), and that
the executable is in your =PATH= (or that you provide the path during execution).
#+name: IqTree2-ManualInstall
#+begin_src sh :exports code :eval never 
tar -xzvf iqtree-2.1.2-Linux.tar.gz
chmod +x iqtree-2.1.2-Linux/bin/iqtree2  # Should not be necessary, but who knows.
mv iqtree-2.1.2-Linux/bin/iqtree2 ~/bin/ # If ~/bin is in your PATH.
#+end_src

* Exercise - Test IQ-TREE2 version

** IQ-TREE2 version                                              :B_alertblock:
:PROPERTIES:
:BEAMER_env: alertblock
:END:
Check that your IQ-TREE2 version agrees with the one I am using.

** Version                                                    :B_ignoreheading:
:PROPERTIES:
:BEAMER_env: ignoreheading
:END:
#+name: IqTree2-Version
#+begin_src sh :exports both :results output verbatim replace
iqtree2 --version
# /path/to/iqtree2 --version
# ./relative/path/to/iqtree2 --version
#+end_src

#+RESULTS: IqTree2-Version
: IQ-TREE multicore version 2.1.2 COVID-edition for Linux 64-bit built Jan  1 1980
: Developed by Bui Quang Minh, James Barbetti, Nguyen Lam Tung,
: Olga Chernomor, Heiko Schmidt, Dominik Schrempf, Michael Woodhams.
: 

* Exercise - Access IQ-TREE2 help

** IQ-TREE2 help                                                 :B_alertblock:
:PROPERTIES:
:BEAMER_env: alertblock
:END:
Access the IQ-TREE2 help, maybe read through some command line flags.

** Help                                                       :B_ignoreheading:
:PROPERTIES:
:BEAMER_env: ignoreheading
:END:
#+name: IqTree2-Help
#+begin_src sh :exports code :eval never
iqtree2 --help | less
#+end_src

#+name: IqTree2-HelpFirstLines
#+begin_src sh :exports results :results verbatim 
iqtree2 --help | head -n 8
#+end_src

#+RESULTS: IqTree2-HelpFirstLines
: IQ-TREE multicore version 2.1.2 COVID-edition for Linux 64-bit built Jan  1 1980
: Developed by Bui Quang Minh, James Barbetti, Nguyen Lam Tung,
: Olga Chernomor, Heiko Schmidt, Dominik Schrempf, Michael Woodhams.
: 
: Usage: iqtree [-s ALIGNMENT] [-p PARTITION] [-m MODEL] [-t TREE] ...
: 
: GENERAL OPTIONS:
:   -h, --help           Print (more) help usages

* Fruit fly data
We are going to analyze some fruit fly data. The data comprises nine
/Drosophila/ populations obtained from [[https://popfly.uab.cat][PopFly]][fn:1].

- NTH :: Netherlands
- EG :: Egypt
- FR :: France
- GA :: Gabon
- GU :: Guinea
- EF :: Ethiopia
- KN :: Kenyia
- SB :: South Africa (Barkly East)
- SP :: South Africa (Phalaborwa)

* Exercise - Explore fruit fly data 

** Explore data                                                  :B_alertblock:
:PROPERTIES:
:BEAMER_env: alertblock
:END:
Have a look at the data in the =./data= folder.

** Counts files and FASTA files                               :B_ignoreheading:
:PROPERTIES:
:BEAMER_env: ignoreheading
:END:
#+name: Data-Explore
#+begin_src sh :exports results :results output 
ls ./data
#+end_src

#+RESULTS: Data-Explore
: data_description.csv
: fruit_flies_10000.cf
: fruit_flies_10000.consensus.fasta
: fruit_flies_10000.random.fasta
: fruit_flies_1000.cf
: fruit_flies_1000.consensus.fasta
: fruit_flies_1000.random.fasta

We have data of two different lengths (1k and 10k sites), and also in counts
file and FASTA file formats.

* Exercise - Run DNA substitution model
Before running PoMo, we will use a normal DNA substitution model.

** DNA substitution model                                        :B_alertblock:
:PROPERTIES:
:BEAMER_env: alertblock
:END:
- Infer a phylogenetic tree using a DNA substitution model.
- Explore the output files. Specifically have a look at the =.log=, the
  =.iqtree= and the =.treefile= files.

** Run DNA substitution model                                 :B_ignoreheading:
:PROPERTIES:
:BEAMER_env: ignoreheading
:END:
#+name: IqTree2-DNASubstitutionModel
#+begin_src sh :exports code :eval never 
iqtree2 -nt 4 -redo -mredo -s fruit_flies_10000.consensus.fasta -B 1000
#+end_src

** Questions                                                   :B_exampleblock:
:PROPERTIES:
:BEAMER_env: exampleblock
:END:
- Which substitution model was used? How was it determined?
- What is the determined maximum log likelihood?
- How does the tree look like (topology, bootstrap values, branch lengths)?

* Exercise - Run PoMo I
** PoMo                                                          :B_alertblock:
:PROPERTIES:
:BEAMER_env: alertblock
:END:
- Infer a phylogenetic tree using PoMo.
- Explore the output files.

** Run PoMO                                                   :B_ignoreheading:
:PROPERTIES:
:BEAMER_env: ignoreheading
:END:
#+name: IqTree2-PoMo
#+begin_src sh :exports code :eval never 
iqtree2 -nt 4 -redo -s fruit_flies_1000.cf -m HKY+F+P -B 1000
#+end_src

** Questions                                                   :B_exampleblock:
:PROPERTIES:
:BEAMER_env: exampleblock
:END:
- What is the average number of samples per population? What is the estimated
  heterozygosity? /Why is it important to check the heterozygosity?/
- Which virtual population size was used?
- What is the determined maximum log likelihood?
- How does the tree look like (topology, bootstrap values, branch lengths)?

* Exercise - Run PoMo II
** PoMo parameters                                               :B_alertblock:
:PROPERTIES:
:BEAMER_env: alertblock
:END:
- Play around with different virtual population sizes.
- Use gamma rate heterogeneity.
- Compare different DNA substitution models.

** Run PoMo with different parameters                         :B_ignoreheading:
:PROPERTIES:
:BEAMER_env: ignoreheading
:END:
#+name: IqTree2-PoMoTilde
#+begin_src sh :exports code :eval never 
f=fruit_flies_1000.cf; m="HKY+F+P+N09+G2"; iqtree2 -nt 4 -redo -s $f -m $m -pre $f.$m
#+end_src

** Questions                                                   :B_exampleblock:
:PROPERTIES:
:BEAMER_env: exampleblock
:END:
- What are the absolute and relative differences between branch lengths for
  different virtual population sizes?
- How do your results compare to using normal DNA substitution models. For a
  fair comparison, you have to run PoMo on the proper data with 10k sites.

* Some results
Inference results are provided in the =./results= folder.

#+name: BranchLengths
#+begin_src ipython :ipyfile PoMo-Workshop-BranchLengths.svg :exports none :results none
cons = 0.7599011228524519
rand = 0.8775802706667841
ns = [3, 5, 7, 9, 15, 19]
pomo = [0.8314274618935616,
        0.7883771094980733,
        0.761713398934836,
        0.7412203731641129,
        0.7004304607585186,
        0.6781954389006495]

plt.plot(1, cons, 'o', label="Consensus")
plt.plot(1, rand, 'o', label="Random")
plt.plot(ns, pomo, 'o', label="PoMo")
plt.legend()
plt.xlabel("Virtual population size")
plt.xticks([1,3,5,7,9,15,19])
plt.xlim([0,20])
plt.ylabel("External / total branch length")
plt.ylim([0.6,1])
#+end_src

#+attr_latex: :width 0.6\textwidth :placement [tb]
#+label: fig-BranchLengths
[[file:PoMo-Workshop-BranchLengths.svg]]

* Advice
Sometimes, the inference is unsuccessful. This may have several reasons. Two of
them are:
- The likelihood derivative is zero or close to zero and numerical underflow
  occurs. This is especially an issue when \(N\) is large. Try using =-safe=
  (which is slower).
- The algorithm diverges. Try repeating the analysis with a different seed.
#+beamer: \vspace{2ex}

In general, it is recommended to *perform replicate analyses* and compare the
parameters and log likelihoods. Further, a *good starting tree* can save a lot
of time and worries.

* Literature
:PROPERTIES:
:ID:       a3a4fe49-79a8-4618-bcae-655485ac54c4
:END:
#+attr_latex: :options [Advanced models with PoMo]
- PoMo :: textcite:DeMaio2015.
- Reversible PoMo :: textcite:Schrempf2016.
- Non-reversible PoMo :: textcite:Schrempf2017.
- Advanced models with PoMo :: textcite:Schrempf2019.
- IQ-TREE2 :: textcite:Minh2020a.
- Consistency of PoMo :: textcite:Borges2020.
- PoMo with selection :: textcite:Borges2019.

* Bibliography
:PROPERTIES:
:BEAMER_opt: allowframebreaks
:END:

#+begin_export latex
\printbibliography{}
#+end_export

* Footnotes

[fn:1] textcite:Hervas2017; thanks Rui for providing the counts files.
